timemachine/cpp/src/potential.cu

```diff
@@ -23,66 +25,63 @@ void Potential::execute_host(
+    DeviceBuffer<double> d_x(N * D);
+    DeviceBuffer<double> d_p(P);
+    DeviceBuffer<double> d_box(D * D);
 
-    gpuErrchk(cudaMalloc(&d_box, D * D * sizeof(double)));
-    gpuErrchk(cudaMemcpy(d_box, h_box, D * D * sizeof(double), cudaMemcpyHostToDevice));
+    gpuErrchk(cudaMemcpy(d_x.data, h_x, d_x.size, cudaMemcpyHostToDevice));
+    gpuErrchk(cudaMemcpy(d_p.data, h_p, d_p.size, cudaMemcpyHostToDevice));
+    gpuErrchk(cudaMemcpy(d_box.data, h_box, d_box.size, cudaMemcpyHostToDevice));
 
-    unsigned long long *d_du_dx = nullptr;
-    unsigned long long *d_du_dp = nullptr;
-    unsigned long long *d_du_dl = nullptr;
-    unsigned long long *d_u = nullptr;
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dx;
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dp;
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dl;
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_u;
 
     // very important that these are initialized to zero since the kernels themselves just accumulate
     if (h_du_dx) {
-        gpuErrchk(cudaMalloc(&d_du_dx, N * D * sizeof(unsigned long long)));
-        gpuErrchk(cudaMemset(d_du_dx, 0, N * D * sizeof(unsigned long long)));
+        d_du_dx.reset(new DeviceBuffer<unsigned long long>(N * D));
+        gpuErrchk(cudaMemset(d_du_dx->data, 0, d_du_dx->size));
     }
     if (h_du_dp) {
-        gpuErrchk(cudaMalloc(&d_du_dp, P * sizeof(unsigned long long)));
-        gpuErrchk(cudaMemset(d_du_dp, 0, P * sizeof(unsigned long long)));
+        d_du_dp.reset(new DeviceBuffer<unsigned long long>(P));
+        gpuErrchk(cudaMemset(d_du_dp->data, 0, d_du_dp->size));
     }
     if (h_du_dl) {
-        gpuErrchk(cudaMalloc(&d_du_dl, N * sizeof(*d_du_dl)));
-        gpuErrchk(cudaMemset(d_du_dl, 0, N * sizeof(*d_du_dl)));
+        d_du_dl.reset(new DeviceBuffer<unsigned long long>(N));
+        gpuErrchk(cudaMemset(d_du_dl->data, 0, d_du_dl->size));
     }
     if (h_u) {
-        gpuErrchk(cudaMalloc(&d_u, N * sizeof(*d_u)));
-        gpuErrchk(cudaMemset(d_u, 0, N * sizeof(*d_u)));
+        d_u.reset(new DeviceBuffer<unsigned long long>(N));
+        gpuErrchk(cudaMemset(d_u->data, 0, d_u->size));
     }
 
-    this->execute_device(N, P, d_x, d_p, d_box, lambda, d_du_dx, d_du_dp, d_du_dl, d_u, static_cast<cudaStream_t>(0));
+    this->execute_device(
+        N,
+        P,
+        d_x.data,
+        d_p.data,
+        d_box.data,
+        lambda,
+        d_du_dx ? d_du_dx->data : nullptr,
+        d_du_dp ? d_du_dp->data : nullptr,
+        d_du_dl ? d_du_dl->data : nullptr,
+        d_u ? d_u->data : nullptr,
+        static_cast<cudaStream_t>(0));
 
     // outputs
     if (h_du_dx) {
-        gpuErrchk(cudaMemcpy(h_du_dx, d_du_dx, N * D * sizeof(*h_du_dx), cudaMemcpyDeviceToHost));
-        gpuErrchk(cudaFree(d_du_dx));
+        gpuErrchk(cudaMemcpy(h_du_dx, d_du_dx->data, d_du_dx->size, cudaMemcpyDeviceToHost));
```

Comment:
Would it be useful to move this line into a `DeviceBuffer::to_host` method or similar?

Reply:
Added in 8d982bf